### PR TITLE
Fix/termination signal handling, fixes #331

### LINKED ISSUE #331: AutoSpotTerminateEvent enters CREATE_FAILED outside us-east-1
# Github issue #

AutoSpotting is failing to create all its components in my AWS account.

The "delegate" AutoSpotTerminationEvent stack fails to create in all regions apart from us-east-1.

Our active region is us-west-2, so I really need AutoSpotting to work there.

## Issue type ##

- Bug Report

## Build number #

```
nightly-1010
```

## Configuration ##

To install AutoSpotting, I used the command from the [Getting Started] guide.

[Getting Started]: https://github.com/AutoSpotting/AutoSpotting/blob/master/START.md

The added profile and region parameters are the only difference.

```
aws cloudformation create-stack \
--stack-name AutoSpotting \
--template-url https://s3.amazonaws.com/cloudprowess/nightly/template.yaml \
--capabilities CAPABILITY_IAM \
--profile dev \
--region us-east-1
```

## Environment ##

- AWS region: all except us-east-1
- Type of environment: DefaultVPC
- Anonymized launch configuration: Not applicable

## Steps to reproduce ##

I ran the command in the Configuration section to install AutoSpotting.

I waited a few minutes for all the stacks to finish creating.

Then I checked the events of the AutoSpotTerminationEvent stack in us-west-2 to reveal the error message.

## Expected results ##

That the AutoSpotTerminationEvent stacks in all regions would enter the CREATE_COMPLETE state.

## Actual results ##

In all regions outside us-east-1, the AutoSpotTerminationEvent stack enters the CREATE_FAILED state.

The LambdaPermissionAutoSpotTerminationNotificationTopic resource gives an error like this:

```
Functions from 'us-east-1' are not reachable in this region ('us-west-2') (Service: AWSLambda; Status Code: 404; Error Code: ResourceNotFoundException; Request ID: 100e219e-4a63-11e9-a748-ef5a0e264c08)
```

### PR DESCRIPTION
**Pull request by cristim**
# Issue Type

- Bugfix Pull Request

## Summary
This should fix #331, you can see more context there.

## Code contribution checklist

1. [x] The contribution fixes a single existing github issue, and it is linked
   to it.
1. [x] The code is as simple as possible, readable and follows the idiomatic Go
  [guidelines](https://golang.org/doc/effective_go.html).
1. [N/A] All new functionality is covered by automated test cases so the overall
  test coverage doesn't decrease.
1. [x] No issues are reported when running `make full-test`.
1. [N/A] Functionality not applicable to all users should be configurable.
1. [N/A] Configurations should be exposed through Lambda function environment
   variables which are also passed as parameters to the
   [CloudFormation](https://github.com/cristim/autospotting/blob/master/cloudformation/stacks/AutoSpotting/template.yaml)
   and
   [Terraform](https://github.com/autospotting/terraform-aws-autospotting/main.tf)
   stacks defined as infrastructure code.
1. [N/A] Global configurations set from the infrastructure stack level should also
   support per-group overrides using tags.
1. [N/A] Tags names and expected values should be similar to the other existing
   configurations.
1. [N/A] Both global and tag-based configuration mechanisms should be tested and
  proven to work using log output from various test runs.
1. [x] The logs should be kept as clean as possible (use log levels as
  appropriate) and formatted consistently to the existing log output.
1. [x] The documentation is updated to cover the new behavior, as well as the
   new configuration options for both stack parameters and tag overrides.
1. [x] A code reviewer reproduced the problem and can confirm the code
  contribution actually resolves it.

### CODE REVIEW

**Comment by mello7tre** [comment]
regional_template.yaml line 43:

`Payload=json.dumps(event),` -> `Payload=dumps(event),`

cause you are directly importing dumps methods: `from json import dumps`

**Comment by cristim** [comment]
Good catch, thanks!

I'll look into this again later today

**Comment by mello7tre** [comment]
in regional_stack_lambda.py - method delete_stack:
waiter.wait - should be better to specify "MaxAttempts" too.

Default for MaxAttempts is 120, with a Delay of 5 seconds (interval between checking status) we have a total max time of 600 sec! That is over the lambda total timeout (5 min).

If there is a problem we are unable to send back signal, cause lambda timeout.

**Comment by cristim** [comment]
I'd rather increase the timeouts to the maximum allowed, 900s.

File: TECHNICAL_DETAILS.md [hunk 1]
```diff
@@ -12,12 +12,13 @@
 | Bid at a certain percentage of the on-demand price | :white_check_mark: (default: 100%) | :white_check_mark: |
 | Can bid the current spot price plus a certain percentage | :white_check_mark: | :white_check_mark: |
 | Automatically determine the cheapest compatible instance type | :white_check_mark: (default) | :white_check_mark: |
-| Can restrict to the same instance type only | :white_check_mark: | :white_check_mark: |
+| Can restrict to the same instance type only | :white_check_mark: - use `current` for the allowed instance types | :white_check_mark: - use `current` for the allowed instance types |
```
[the remark this answers is not in the document]
Comment (cristim) [reply TECHNICAL_DETAILS.md]: all these long lines are marked as wontfix, we actually ignore them using markdownint tags but the codeclimate scanner doesn't support those at the moment.

No errors are reported when locally executing a proper markdownlint implementation that implements support for these tag overrides

**Comment by cristim** [comment]
I also tested the imminent spot termination handler functionality across regions and it worked as expected.

**Comment by mello7tre** [comment]
I have begin testing it, some notes:
* regional_stack_lambda.py - line 111 (method send):
`response_url = event['response_url']` -> `response_url = event['ResponseURL']` 
* regional_stack_lambda.py - line 53 (method delete_stack):
should be better to execute cfn.delete_stack in a try block
* regional_template.yaml - as role are cross region - there is no need to create a LambdaExecutionRole and relative LambdaPolicy in all regions. We can just create one in the main template and pass it as parameter to the lambda that create the regional ones.
* regional_template.yaml there is no need to use sns at all, regional CW events can send directly to regional lambda (if we do not want change code in autospotting.go we can "encapsulate" the events in an sns message:
```                
                snsEvent = {
                    'Records': [
                        {   
                            'Event': 'aws:sns',
                            'EventSource': '1.0',
                            'EventSubscriptionArn': '',
                            'Sns': {
                                'Type': 'Notification',
                                'Message': event,
                            }
                        }
                    ]
                }

```

**Comment by cristim** [comment]
Please add your comments directly on the code, it makes it easier to answer or explain if I need to.

> I have begin testing it, some notes:
> 
> * regional_stack_lambda.py - line 111 (method send):
>   `response_url = event['response_url']` -> `response_url = event['ResponseURL']`
Good catch, fixed in the next iteration

> * regional_stack_lambda.py - line 53 (method delete_stack):
>   should be better to execute cfn.delete_stack in a try block
Actually we use try when we call this function in create_stack, so that we can ignore the case when the stack doesn't exist and we fail to delete it. But exceptions when deleting the stack on all regions are valid IMHO.

> * regional_template.yaml - as role are cross region - there is no need to create a 
> LambdaExecutionRole and relative LambdaPolicy in all regions. We can just create one in the main 
> template and pass it as parameter to the lambda that create the regional ones.
> * regional_template.yaml there is no need to use sns at all, regional CW events can send directly to regional lambda (if we do not want change code in autospotting.go we can "encapsulate" the events in an sns message:
> 
> ```
>                 snsEvent = {
>                     'Records': [
>                         {   
>                             'Event': 'aws:sns',
>                             'EventSource': '1.0',
>                             'EventSubscriptionArn': '',
>                             'Sns': {
>                                 'Type': 'Notification',
>                                 'Message': event,
>                             }
>                         }
>                     ]
>                 }
> ```
I would prefer to ship this fix as it is ASAP and maybe change this later. Feel free to submit another PR with these further simplifications.

**Comment by mello7tre** [comment]
One question:
in regional_stack_lambda.py, you have replaced `cfnresponse.send(event, context, cfnresponse.SUCCESS, {})`
with a more long code that create a response (method send).
If i understand correctly you have done this to be able to specify a custom " Reason":
```
    response_body['Reason'] = \
        'See the details in CloudWatch Log Stream: ' + context.log_stream_name
```
but the only place where i can see that message is in the log stream itself!

But maybe the reason is another and i am missing something...

**Comment by cristim** [comment]
It turns out the `cfnresponse` module is not available if we ship the Lambda function's Python code as a zip file, it's only available when embedding it inline in the CloudFormation template. I wanted the file to be stand-alone so I can get editor support when working on it, since it was quite large, and also it was making the CloudFormation template huge.

That's the reason why I had to implement the send method as documented at 
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-code.html#cfn-lambda-function-code-cfnresponsemodule

**Comment by mello7tre** [comment]
Ok thanks for the explanation.
About the template simplifications, i have already a working/tested version, but if you prefer i will wait that you merge this PR and then i will submit another one.

**Comment by mello7tre** [comment]
Found a problem when deleting stacks:
RegionalStackCreationLambdaFunction and RegionalStackCustomResource have a `DependsOn: LambdaPolicy` but i think the right one should be on `LambdaRegionalStackPolicy`.

While deleting, LambdaRegionalStackPolicy and LambdaRegionalStackExecutionRole are deleted before signaling RegionalStackCustomResource so when RegionalStackCreationLambdaFunction try to delete regional stack, AutoSpotTeminationEventRule goes in error with:
```
User: arn:aws:sts::429548892436:assumed-role/t-spot-LambdaRegionalStackExecutionRole-13G3108DXOHQ6/t-spot-RegionalStackCreationLambdaFunction-1GFI9CBJASOX2 is not authorized to perform: events:RemoveTargets on resource: arn:aws:events:eu-west-1:429548892436:rule/AutoSpottingRegionalResou-AutoSpotTeminationEventR-ZPY8F8UAN4L9
```

**Comment by cristim** [comment]
@mello7tre I'd like to merge this, unless you have any other concerns.

**Comment by mello7tre** [comment]
Ok go on, once merged i will open a new draft PR for the template simplifications.

**Pull request merged**